src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,183 @@
+        dai = DSTokenAbstract(_dai);
+        gem = DSTokenAbstract(_gem);
+        psm = PsmAbstract(_psm);
+        to = _to;
+
+        require(GemJoinAbstract(psm.gemJoin()).gem() == _gem, "RwaInputConduit3/wrong-gem-for-psm");
+        require(address(psm.dai()) == _dai, "RwaInputConduit3/wrong-dai-for-psm");
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "RwaInputConduit3/not-authorized");
+        _;
+    }
+
+    /**
+     * @notice Grants `usr` admin access to this contract.
+     * @param usr The user address.
+     */
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` admin access from this contract.
+     * @param usr The user address.
+     */
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    /**
+     * @notice Grants `usr` push access to this contract.
+     * @param usr The user address.
+     */
+    function mate(address usr) external auth {
+        may[usr] = 1;
+        emit Mate(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` push access from this contract.
+     * @param usr The user address.
+     */
+    function hate(address usr) external auth {
+        may[usr] = 0;
+        emit Hate(usr);
+    }
+
+    /**
+     * @notice Internal method which:
+     *  - Approve PSM gemJoin Adapter with WAD amount of GEM
+     *  - Swap GEM to DAI through PSM
+     *  - Push DAI to Recipient address (to)
+     * @param wad Amount to swap and push
+     */
+    function _swapAndPush(uint256 wad) internal {
+        gem.approve(address(psm.gemJoin()), wad);
+        psm.sellGem(address(this), wad);
+
+        uint256 balance = dai.balanceOf(address(this));
+        dai.transfer(to, balance);
+
+        emit Push(to, balance);
+    }
+
+    /**
+     * @notice Method to swap WAD amount of USDC contract balance to DAI through PSM and push it to Recipient address (to).
+     * @dev `msg.sender` must first receive push acess through mate().
+     * @param wad Amount to swap and push
+     */
+     function push(uint256 wad) public {
+        require(may[msg.sender] == 1, "RwaInputConduit3/not-mate");
+        require(wad > 0, "RwaInputConduit3/not-positive-amount");
+        uint256 balance = gem.balanceOf(address(this));
+        require(balance >= wad, "RwaInputConduit3/insufficient-gem-balance");
+
+        _swapAndPush(wad);
+    }
+
+    /**
+     * @notice Method to swap USDC contract balance to DAI through PSM and push it into RwaUrn address.
+     * @dev `msg.sender` must first receive push acess through mate().
+     */
+     function push() public {
+        require(may[msg.sender] == 1, "RwaInputConduit3/not-mate");
+        uint256 balance = gem.balanceOf(address(this));
+        require(balance > 0, "RwaInputConduit3/insufficient-gem-balance");
```

Comment:
We probably don't need this guard here.
If someone calls `psm.sellGem` with `0`, it will simply be a no-op.

Reply:
I put it there just to save some gas

Reply:
> Looks good @nduchak subject to changes proposed by @hbarcelos - external to this and maybe you discussed this yesterday - @lollike - what about edge cases if they start charging fees or implementing negative rates - see this [makerdao/dss-psm#6](https://github.com/makerdao/dss-psm/pull/6) maybe something that doesn't need to be addressed in the conduit- but in the architecture

Another edge case which we can have is DC of PSM, if we try to swap amount above DC then it will fail and our USDC will be stacked in conduit. Does it make sense to have a method to move USDC out from conduit???

Reply:
> I put it there just to save some gas, as it will fail on this line: https://github.com/makerdao/dss-psm/blob/master/src/psm.sol#L113

Why would it fail in that line?

It's actually:
```
gemJoin.join(address(this), gemAmt, msg.sender);
```

For USDC we use [join-5-auth](https://github.com/makerdao/dss-psm/blob/7d717d88d6b6dfcb346119369647969df6cadad3/src/join-5-auth.sol). There is this require statement on L84:
```
require(int256(wad) >= 0, "AuthGemJoin5/overflow");
```

Reply:
Yes, I mean this one if you go deeper.
We save some gas skipping these operations:
PSM:
```
uint256 gemAmt18 = mul(gemAmt, to18ConversionFactor);
uint256 fee = mul(gemAmt18, tin) / WAD;
uint256 daiAmt = sub(gemAmt18, fee);
```

GemJoin
```
 require(live == 1, "AuthGemJoin5/not-live");
 uint256 wad = mul(amt, 10 ** (18 - dec));
```

Reply:
I see...

Another point is that having this check actually deviates from the behavior of the other conduits, which do not fail when calling `push` with balance `0`:
https://github.com/clio-finance/mip21-toolkit/blob/3fdb932f6daf966223643add1e191397461f7c9a/src/conduits/RwaInputConduit2.sol#L130-L137

Reply:
Hmm... Is it something which we should follow? If yes it's just optimization and we can remove it.

Reply:
Probably this is something we want to keep consistent across implementations. That `push()` method is the primary function of this contract, it shouldn't behave on unexpected ways.